semgrep/semgrep/target_manager.py

```diff
@@ -0,0 +1,234 @@
+from semgrep.error import UnknownLanguageError
+from semgrep.util import partition_set
+
+
+def lang_to_exts(language: str) -> List[str]:
+    """
+        Convert language to expected file extensions
+
+        If language is not a supported semgrep language then
+        raises UnknownLanguageError
+    """
+    if language in ["python", "python2", "python3", "py"]:
+        return ["py", "pyi"]
+    elif language in ["js", "javascript"]:
+        return ["js"]
+    elif language in ["java"]:
+        return ["java"]
+    elif language in ["c"]:
+        return ["c"]
+    elif language in ["go", "golang"]:
+        return ["go"]
+    elif language in ["ml", "ocaml"]:
+        return ["mli", "ml", "mly", "mll"]
+    else:
+        raise UnknownLanguageError(f"Unsupported Language: {language}")
+
+
+class TargetManager:
+    def __init__(
+        self,
+        includes: List[str],
+        excludes: List[str],
+        targets: List[str],
+        visible_to_git_only: bool = False,
+    ) -> None:
+        """
+            Handles all file include/exclude logic for semgrep
+
+            If visible_to_git_only is true then will only consider files that are
+            tracked or (untracked but not ignored) by git
+        """
+        self._targets = targets
+        self._includes = includes
+        self._excludes = excludes
+        self._visible_to_git_only = visible_to_git_only
+
+        self._filtered_targets: Dict[str, Set[Path]] = {}
+
+    @staticmethod
+    def resolve_targets(targets: List[str]) -> Set[Path]:
+        """
+            Return list of Path objects appropriately resolving relative paths
+            (relative to cwd) if necessary
+        """
+        base_path = Path(".")
+        return set(
+            Path(target) if Path(target).is_absolute() else base_path.joinpath(target)
+            for target in targets
+        )
+
+    @staticmethod
+    def _parse_output(output: str, curr_dir: Path) -> Set[Path]:
+        """
+            Convert a newline delimited list of files to a set of path objects
+            prepends curr_dir to all paths in said list
+
+            If list is empty then returns an empty set
+        """
+        files: Set[Path] = set()
+        if output:
+            files = set(Path(curr_dir) / elem for elem in output.strip().split("\n"))
+        return files
+
+    @staticmethod
+    def _expand_dir(
+        curr_dir: Path, language: str, visible_to_git_only: bool
+    ) -> Set[Path]:
+        """
+            Recursively go through a directory and return list of all files with
+            default file extention of language
+        """
+        extensions = lang_to_exts(language)
+        expanded: Set[Path] = set()
+
+        for ext in extensions:
+            if visible_to_git_only:
+                try:
+                    # Tracked files
+                    tracked_output = subprocess.check_output(
+                        ["git", "ls-files", f"*.{ext}"],
```

Comment:
what's the run time of this command? Seems slow compared to extracting globs from .gitignore

Reply:
`find . -type f -name "*.py"  0.00s user 0.00s system 87% cpu 0.007 total`
`git ls-files "*.py"  0.00s user 0.01s system 74% cpu 0.014 total`

Negligible difference with find tbh